### PR TITLE
[Repo Config] Add Packages: glog, gflags and abseil

### PR DESCRIPTION
**Pull request by natsunoyoru97**
### Why we have the PR
- The c++ default iostream is not good to use and the third-party library is in need, here we use `glog` to do this.
- `gflags` is the library to parse the flags of the command.
- `abseil` is a C++ utility library to make code easier to write, e.g. to join the string in a vector.

### What this PR did
- Import three packages by xmake: `glog`, `gflags`, `abseil`

### Testcase
There are three functions altogether that start with "Init", and two of them test whether the packages are successfully imported:

```C++
/// TODO(natsunoyoru97): redirect glog to stderr
void InitGlog(const char* argv0) { google::InitGoogleLogging(argv0); }

void InitGflags() {
  // TODO(natsunoyoru97): Use glog to replace the cout
  std::cout << FLAGS_test << "\n";
}

void InitAbseil() {
  std::vector<std::string> vec{"hello", "world"};
  std::string s = absl::StrJoin(vec, " ");

  // TODO(natsunoyoru97): Use glog to replace the cout
  std::cout << s << "\n";
}
```

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -74,8 +82,36 @@ void ExecuteDb(const std::string& command,
...
+  std::vector<std::string> vec{"hello", "world"};
+  std::string s = absl::StrJoin(vec, " ");
+
+  // TODO(natsunoyoru97): Use glog to replace the cout
+  std::cout << s << "\n";
+}
+
 int main(int argc, char** argv) {
-  std::cout << "Simple database" << std::endl;
+  /* TODO(natsunoyoru97): It seems that the result is not as expected.
```
Comment (dentiny) [inline src/main.cpp:107]: Curious what's the issue?

File: src/main.cpp [hunk 2]
```diff
@@ -74,8 +82,36 @@ void ExecuteDb(const std::string& command,
   }
 }
 
+// The signal handler for SIGPIPE
+void DoNothing(int signum) {
+  std::cout << "SIGPIPE"
+            << "\n"; /* Ignore SIGPIPE */
+}
+
+void InitGlog(const char* argv0) { google::InitGoogleLogging(argv0); }
```
Comment (dentiny) [inline src/main.cpp:91]: I remember we also talked about redirect it to stderr?

File: src/main.cpp (hunk 1 above)
Comment (dentiny) [reply src/main.cpp:107]: I think we should ignore `SIGPIPE`.
```
signal(SIGPIPE, SIG_IGN);
```

File: src/main.cpp [hunk 3]
```diff
@@ -74,8 +82,36 @@ void ExecuteDb(const std::string& command,
...
+
+void InitGflags() {
+  // TODO(natsunoyoru97): Use glog to replace the cout
+  std::cout << FLAGS_test << "\n";
+}
+
+void InitAbseil() {
```
Comment (dentiny) [inline src/main.cpp:98]: Don't think we need the function.

File: xmake.lua [hunk 4]
```diff
@@ -1,11 +1,12 @@
 set_languages("c++17")
 
-add_requires("doctest", "gtest ~1.12.1")
+add_requires("abseil", "doctest", "gtest ~1.12.1", "glog", "gflags")
 add_rules("mode.debug", "mode.release")
 
 target("target")
     set_kind("binary")
     add_files("src/*.cpp")
+    add_packages("glog", "gflags")
```
Comment (dentiny) [inline xmake.lua:9]: curious, abseil is not required in the dependency package and it works?

File: src/main.cpp (hunk 3 above)
Comment (dentiny) [reply src/main.cpp:98]: If it's for testing, your could share a code snippet in the PR description.

**Review by dentiny — COMMENTED** [review]
btw, I somehow think it's better to keep the previous PR.
- There's not too many changes between these two.
- It keeps our conversation.
But it's fine for now.

File: src/main.cpp (hunk 1 above)
Comment (natsunoyoru97) [reply src/main.cpp:107]: Yes I did call this API, but it just DID NOT WORK. I tried SIGINT to test whether it works, but the process still exits even if I raise a SIGINT signal.

File: src/main.cpp (hunk 2 above)
Comment (natsunoyoru97) [reply src/main.cpp:91]: Yes we did. I will hand in another PR to do this.

File: xmake.lua (hunk 4 above)
Comment (natsunoyoru97) [reply xmake.lua:9]: Yes, it works even if I did not add the line.

File: src/main.cpp (hunk 3 above)
Comment (natsunoyoru97) [reply src/main.cpp:98]: > Don't think we need the function.

OK I will drop it.

File: src/main.cpp (hunk 2 above)
Comment (dentiny) [reply src/main.cpp:91]: Leave a TODO here?
Actually it's just one line change.

File: src/main.cpp (hunk 1 above)
Comment (dentiny) [reply src/main.cpp:107]: Could you please tell me how to repro?

Comment (natsunoyoru97) [reply src/main.cpp:107]: The bug is weird and I am trying to repro it. I will upload a screenshot to show its behavior.

Comment (natsunoyoru97) [reply src/main.cpp:100]: 2023-01-01 12-08-43屏幕截图
This is the result after I compile the code and run the binary (here I use SIGINT to test if it works). Even if the handler caught the signal, but the process still exits.

Comment (dentiny) [reply src/main.cpp:100]: btw, could you please paste the *CODE* rather than image so that others could copy & paste?

File: xmake.lua (hunk 4 above)
Comment (dentiny) [reply xmake.lua:9]: I think it doesn't work on my end.
```
error: src/main.cpp:11:10: fatal error: absl/strings/str_join.h: No such file or directory
   11 | #include "absl/strings/str_join.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

Comment (dentiny) [reply xmake.lua:9]: Even if I add it to the dependency, it fails at linking.
```
error: /usr/bin/ld: cannot find -lgtest
collect2: error: ld returned 1 exit status
```

Comment (dentiny) [reply xmake.lua:9]: Could you please check whether you have installed on your system directory?

Comment (dentiny) [reply xmake.lua:9]: Could you please check whether you have installed on your system directory?

File: src/main.cpp (hunk 1 above)
Comment (dentiny) [reply src/main.cpp:100]: btw, I just checked,
```
signal(SIGINT, SIG_IGN);
```
works on my end. Could you please double check, and tell me how you test on you end?

Comment (natsunoyoru97) [reply src/main.cpp:100]:
```c++
#include <csignal>

// The signal handler for SIGPIPE
void DoNothing(int signum) {
  std::cout << "SIGPIPE"
            << "\n"; /* Ignore SIGPIPE */
}

int main(int argc, char** argv) {
  std::signal(SIGINT, DoNothing);
  // Some init functions
   
   while (true) {
    // code to deal with the commands
  }
}
```
This is the code snippet for the test.
I also tried 
```c++
signal(SIGINT, SIG_IGN);
```
but it didn't work. All I did it to run this shell command:
```shell
xmake && xmake run target
```
I am making a double check for the possible reasons for the bug.

File: xmake.lua (hunk 4 above)
Comment (natsunoyoru97) [reply xmake.lua:9]: I did not install abseil at first, and xmake fetched the package from the remote git repository. I'll make a check for this.

Comment (natsunoyoru97) [reply xmake.lua:9]: > error: /usr/bin/ld: cannot find -lgtest

btw, I installed gtest before I wrote the xmake script.

Comment (natsunoyoru97) [reply xmake.lua:9]: > Could you please check whether you have installed on your system directory?

I found that abseil is installed on these directories:
```
/usr/share/bazel/third_party/py/abseil/absl
/usr/include/absl
/usr/lib/cmake/absl
```

Comment (dentiny) [reply xmake.lua:9]: That's what I mean. System directories are usually automatically included in the build dependency.
If you check the compilation/link command, you should be able to see `-I/...` and `-L/...`.
These packages are usually installed by `sudo apt install` or `make install`, please delete them from system directories and retry.

**Comment by dentiny** [comment]
Another feature request: enable C++ compilation options like `-pedantic, -Wall`, etc. Not have to be this PR.

File: src/main.cpp (hunk 1 above)
Comment (dentiny) [reply src/main.cpp:100]: It works on my end.

Comment (dentiny) [reply src/main.cpp:100]: I copied your code snippet, and hit `ctrl + C`.
The program doesn't exit, I think that's what we expect?

**Review by dentiny — COMMENTED** [review]
I still fail on my local.
```
  => install gflags v2.2.2 .. failed

-- Configuring incomplete, errors occurred!
See also "/home/hjiang/.xmake/cache/packages/2301/g/gflags/v2.2.2/source/build_0f545414/CMakeFiles/CMakeOutput.log".
 build tool.
CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
if you want to get more verbose errors, please see:
  -> /home/hjiang/.xmake/cache/packages/2301/g/gflags/v2.2.2/installdir.failed/logs/install.txt
error: install failed!
```

**Comment by natsunoyoru97** [comment]
> I still fail on my local.
> 
> ```
>   => install gflags v2.2.2 .. failed
> 
> -- Configuring incomplete, errors occurred!
> See also "/home/hjiang/.xmake/cache/packages/2301/g/gflags/v2.2.2/source/build_0f545414/CMakeFiles/CMakeOutput.log".
>  build tool.
> CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
> if you want to get more verbose errors, please see:
>   -> /home/hjiang/.xmake/cache/packages/2301/g/gflags/v2.2.2/installdir.failed/logs/install.txt
> error: install failed!
> ```

The code can be built on my local. I will set CMAKE_CXX_COMPILER on the script and make it run on this PR.

**Review by dentiny — APPROVED** [review]
My bad: I switch to a new VM and forgot to install make...
I think it's good to go.

**Pull request merged**